# Data Science Pandas

## Tasks Today:

0) <b>Pre-Work</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Numpy Random Sampling

1) <b>Pandas</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Importing <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Tabular Data Structures <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - from_dict() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - read_csv() <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) <b>In-Class Exercise #1</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Accessing Data <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Indexing <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - df.loc <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - keys() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Slicing a DataFrame <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Built-In Methods <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - head() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - tail() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - shape <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - describe() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - sort_values() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - .columns <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) <b>In-Class Exercise #2</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; g) Filtration <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Conditionals <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Subsetting <br>
 &nbsp;&nbsp;&nbsp;&nbsp; h) Column Transformations <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Generating a New Column w/Data <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - User Defined Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; i) Aggregations <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - groupby() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Type of groupby() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - mean() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - groupby() w/Multiple Columns <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - drop_duplicates() <br>

## Numpy Random Sampling

In [3]:
import numpy as np

# A single call generates a single random number
print('Here is a random number: %s' % np.random.uniform())
print(f'Here is a random number: {np.random.uniform()}')

# You can also pass some bounds
print('Here is a random number between 0 and 1 Million: %s' % np.random.uniform(0, 1e6))

# You can also generate a bunch of random numbers all at once
print('Here is a random number between 0 and 1 million: %s' % np.random.uniform(0, 1e6, 3))
      
# Even matrices and tensors with weird shapes
print('Here is a 3x3 matrix of random numbers between 0 and 1 Million: \n %s' % np.random.uniform(0, 1e6, (3, 3)))

# instead of float values, generating random integers
print('Using Random Integer: %s' % np.random.randint(0, 10, 4))

Here is a random number: 0.8674685945964689
Here is a random number: 0.042581141282167945
Here is a random number between 0 and 1 Million: 207439.5206167452
Here is a random number between 0 and 1 million: [658207.93392022 258494.05314398 869785.28001131]
Here is a 3x3 matrix of random numbers between 0 and 1 Million: 
 [[660855.54764296 304338.87166967 819220.66437627]
 [ 91915.89869386 370468.16430838 793377.49394816]
 [864151.83369709 384760.81162928 459166.47111184]]
Using Random Integer: [7 0 1 3]


## Pandas <br>

<p>Pandas is a flexible data analysis library built on top of NumPy that is excellent for working with tabular data. It is currently the de-facto standard for Python-based data analysis, and fluency in Pandas will do wonders for your productivity and frankly your resume. It is one of the fastest ways of getting from zero to answer in existence. </p>

<ul>
    <li>Pandas is a Python module, written in C. The Pandas module is a high performance, highly efficient, and high level data analysis library. It allows us to work with large sets of data called dataframes.</li>
    <li>Series is a one-dimensional labeled array capable of holding data of any type (integer, string, float, python objects, etc.)</li>
    <li>Dataframe = Spreadsheet (has column headers, index, etc.)</li>
</ul>

### Importing

In [3]:
!pip install pandas

In [4]:
import pandas as pd

# always use pd, standard for data science

### Tabular data structures <br>
<p>The central object of study in Pandas is the DataFrame, which is a tabular data structure with rows and columns like an excel spreadsheet. The first point of discussion is the creation of dataframes both from native Python dictionaries, and text files through the Pandas I/O system.</p>

In [5]:
names = ['Alice',
        'Bob',
        'James',
        'Beth',
        'John',
        'Sally',
        'Richard',
        'Lauren',
        'Brandon',
        'Sabrina']

ages = np.random.randint(18,35,len(names))

my_people = {
    'names': names,
    'ages': ages
}

my_people

{'names': ['Alice',
  'Bob',
  'James',
  'Beth',
  'John',
  'Sally',
  'Richard',
  'Lauren',
  'Brandon',
  'Sabrina'],
 'ages': array([19, 31, 28, 18, 30, 34, 22, 19, 20, 33])}

##### from_dict()

<p>Let's convert our not-so-useful-for-analysis dict into a Pandas dataframe. We can use the from_dict function to do this easily using Pandas:</p>

In [6]:
data = pd.DataFrame.from_dict(my_people)

data

names  ages
0    Alice    19
1      Bob    31
2    James    28
3     Beth    18
4     John    30
5    Sally    34
6  Richard    22
7   Lauren    19
8  Brandon    20
9  Sabrina    33

##### read_csv()

In [7]:
# pd.set_option('display.max_columns', None)
# pd.set_option('display.max_rows', None)

marathon = pd.read_csv('../Downloads/boston_marathon2017_edited.csv', sep = ',')
marathon

10K                     15K                    20K  \
0      12/30/1899 12:30:28 AM  12/30/1899 12:45:44 AM  12/30/1899 1:01:15 AM   
1      12/30/1899 12:30:27 AM  12/30/1899 12:45:44 AM  12/30/1899 1:01:15 AM   
2      12/30/1899 12:30:29 AM  12/30/1899 12:45:44 AM  12/30/1899 1:01:16 AM   
3      12/30/1899 12:30:29 AM  12/30/1899 12:45:44 AM  12/30/1899 1:01:19 AM   
4      12/30/1899 12:30:28 AM  12/30/1899 12:45:44 AM  12/30/1899 1:01:15 AM   
...                       ...                     ...                    ...   
26405   12/30/1899 1:35:41 AM   12/30/1899 2:23:35 AM  12/30/1899 3:12:44 AM   
26406   12/30/1899 1:05:33 AM   12/30/1899 1:52:17 AM  12/30/1899 2:49:41 AM   
26407   12/30/1899 1:43:36 AM   12/30/1899 2:32:36 AM                    NaN   
26408   12/30/1899 1:27:19 AM   12/30/1899 2:17:17 AM  12/30/1899 3:11:40 AM   
26409   12/30/1899 1:17:12 AM   12/30/1899 2:00:10 AM  12/30/1899 2:58:55 AM   

                         25K                    30K                    35K  \
0      12/30/1899 1:16:59 AM  12/30/1899 1:33:01 AM  12/30/1899 1:48:19 AM   
1      12/30/1899 1:16:59 AM  12/30/1899 1:33:01 AM  12/30/1899 1:48:19 AM   
2      12/30/1899 1:17:00 AM  12/30/1899 1:33:01 AM  12/30/1899 1:48:31 AM   
3      12/30/1899 1:17:00 AM  12/30/1899 1:33:01 AM  12/30/1899 1:48:58 AM   
4      12/30/1899 1:16:59 AM  12/30/1899 1:33:01 AM  12/30/1899 1:48:41 AM   
...                      ...                    ...                    ...   
26405  12/30/1899 4:12:06 AM  12/30/1899 5:03:08 AM  12/30/1899 5:55:18 AM   
26406  12/30/1899 3:50:19 AM  12/30/1899 4:50:01 AM  12/30/1899 5:53:48 AM   
26407  12/30/1899 4:15:21 AM  12/30/1899 5:06:37 AM  12/30/1899 6:00:33 AM   
26408  12/30/1899 4:06:10 AM  12/30/1899 5:07:09 AM  12/30/1899 6:06:07 AM   
26409  12/30/1899 4:27:14 AM  12/30/1899 5:37:13 AM  12/30/1899 6:39:07 AM   

                         40K                      5K  Age      Bib  ...  \
0      12/30/1899 2:02:53 AM  12/30/1899 12:15:25 AM   24     11.0  ...   
1      12/30/1899 2:03:14 AM  12/30/1899 12:15:24 AM   30     17.0  ...   
2      12/30/1899 2:03:38 AM  12/30/1899 12:15:25 AM   25     23.0  ...   
3      12/30/1899 2:04:35 AM  12/30/1899 12:15:25 AM   32     21.0  ...   
4      12/30/1899 2:05:00 AM  12/30/1899 12:15:25 AM   31      9.0  ...   
...                      ...                     ...  ...      ...  ...   
26405  12/30/1899 6:46:57 AM  12/30/1899 12:46:44 AM   61  25166.0  ...   
26406  12/30/1899 6:54:21 AM  12/30/1899 12:32:03 AM   25  25178.0  ...   
26407  12/30/1899 6:54:38 AM  12/30/1899 12:53:11 AM   57  27086.0  ...   
26408  12/30/1899 6:56:08 AM  12/30/1899 12:40:34 AM   64  25268.0  ...   
26409  12/30/1899 7:41:23 AM  12/30/1899 12:39:36 AM   48  25266.0  ...   

       Division Gender                   Half  M/F                      Name  \
0             1      1  12/30/1899 1:04:35 AM    M           Kirui, Geoffrey   
1             2      2  12/30/1899 1:04:35 AM    M               Rupp, Galen   
2             3      3  12/30/1899 1:04:36 AM    M             Osako, Suguru   
3             4      4  12/30/1899 1:04:45 AM    M          Biwott, Shadrack   
4             5      5  12/30/1899 1:04:35 AM    M            Chebet, Wilson   
...         ...    ...                    ...  ...                       ...   
26405       344  11972  12/30/1899 3:23:31 AM    F  Steinbach, Paula Eyvonne   
26406      4774  14436  12/30/1899 3:00:26 AM    M        Avelino, Andrew R.   
26407       698  11973  12/30/1899 3:36:24 AM    F           Hantel, Johanna   
26408      1043  14437  12/30/1899 3:22:30 AM    M              Reilly, Bill   
26409      2553  14438  12/30/1899 3:08:16 AM    M             Rigsby, Scott   

      Number of Records          Official Time  Overall  \
0                     1  12/30/1899 2:09:37 AM        1   
1                     1  12/30/1899 2:09:58 AM        2   
2                     1  12/30/1899 2:10:28 AM        3   
3                     1  12/30

### In-Class Exercise #1 - Read in Boston Red Sox Hitting Data <br>
<p>Use the pandas read_csv() method to read in the statistics from the two files yesterday.</p>

In [8]:
import pandas as pd

bos17 = pd.read_csv('../Downloads/redsox_2017_hitting.txt', sep = ',')
bos18 = pd.read_csv('../Downloads/redsox_2018_hitting.txt', sep = ',')

bos17

Rk Pos                Name  Age    G   PA   AB    R    H  2B  ...    OBP  \
0    1   C   Christian Vazquez   26   99  345  324   43   94  18  ...  0.330   
1    2  1B      Mitch Moreland   31  149  576  508   73  125  34  ...  0.326   
2    3  2B      Dustin Pedroia   33  105  463  406   46  119  19  ...  0.369   
3    4  SS     Xander Bogaerts   24  148  635  571   94  156  32  ...  0.343   
4    5  3B       Rafael Devers   20   58  240  222   34   63  14  ...  0.338   
5    6  LF   Andrew Benintendi   22  151  658  573   84  155  26  ...  0.352   
6    7  CF  Jackie Bradley Jr.   27  133  541  482   58  118  19  ...  0.323   
7    8  RF        Mookie Betts   24  153  712  628  101  166  46  ...  0.344   
8    9  DH      Hanley Ramirez   33  133  553  496   58  120  24  ...  0.320   
9   10   C          Sandy Leon   28   85  301  271   32   61  14  ...  0.290   
10  11  UT         Chris Young   33   90  276  243   30   57  12  ...  0.322   
11  12  3B       Deven Marrero   26   71  188  171   32   36   9  ...  0.259   
12  13  2B       Eduardo Nunez   30   38  173  165   23   53  12  ...  0.353   
13  14  2B          Brock Holt   29   64  164  140   20   28   6  ...  0.305   
14  15  IF       Josh Rutledge   28   37  118  107   10   24   2  ...  0.297   
15  16  3B      Pablo Sandoval   30   32  108   99   10   21   2  ...  0.269   
16  17  UT          Sam Travis   23   33   83   76   13   20   6  ...  0.325   
17  18  UT         Tzu-Wei Lin   23   25   66   56    7   15   0  ...  0.369   
18  19  IF     Marco Hernandez   24   21   60   58    7   16   3  ...  0.300   
19  20  UT         Rajai Davis   36   17   38   36    7    9   2  ...  0.289   
20  21  UT        Steve Selsky   27    8    9    9    0    1   1  ...  0.111   
21  22  UT       Blake Swihart   25    6    7    5    1    1   0  ...  0.429   
22  23  2B      Chase d'Arnaud   30    2    1    1    2    1   0  ...  1.000   

      SLG    OPS  OPS+   TB  GDP  HBP  SH  SF  IBB  
0   0.404  0.735    91  131   14    3   0   1    0  
1   0.443  0.769    99  225   14    6   0   5    6  
2   0.392  0.760   100  159   11    2   2   4    4  
3   0.403  0.746    95  230   17    6   0   2    6  
4   0.482  0.819   111  107    5    0   0   0    3  
5   0.424  0.776   102  243   16    6   1   8    7  
6   0.402  0.726    89  194    8    9   0   2    4  
7   0.459  0.803   108  288    9    2   0   5    9  
8   0.429  0.750    94  213   15    6   0   0    8  
9   0.354  0.644    68   96    5    1   1   3    1  
10  0.387  0.709    85   94    4    2   0   1    0  
11  0.333  0.593    54   57    8    0   3   2    0  
12  0.539  0.892   128   89    3    2   0   0    0  
13  0.243  0.548    47   34    3    3   0   2    0  
14  0.262  0.558    49   28    1    2   0   0    0  
15  0.354  0.622    61   35    4    0   0   1    0  
16  0.342  0.667    75   26    2    1   0   0    0  
17  0.339  0.709    88   19    0    0   1   0    0  
18  0.328  0.628    65   19    0    1   0   0    0  
19  0.306  0.595    56   11    2    1   0   0    0  
20  0.222  0.333   -16    2    0    0   0   0    0  
21  0.200  0.629    74    1    0    0   0   0    0  
22  1.000  2.000   428    1    0    0   0   0    0  

[23 rows x 28 columns]

### Accessing Data <br>

##### Indexing

<p>You can directly select a column of a dataframe just like you would a dict. The result is a Pandas 'Series' object.</p>

In [9]:
data_ages = data['ages']

print(data_ages)

# Print out the data type of the variable data

# Series
print(type(data_ages))
# DataFrame
print(type(data))

# DataSeries Index - starting at 0
print(data_ages[3])

# key first, row index second
print(data['ages'][3])

# print(float(data_ages[0]))
# print(data_ages)

0    19
1    31
2    28
3    18
4    30
5    34
6    22
7    19
8    20
9    33
Name: ages, dtype: int32
<class 'pandas.core.series.Series'>
<class 'pandas.core.frame.DataFrame'>
18
18


##### df.loc

<p>Along the horizontal dimension, rows of Pandas DataFrames are Row objects. You will notice there is a third column present in the DataFrame - this is the $\textit{index}$. It is automatically generated as a row number, but can be reassigned to a column of your choice using the DataFrame.set_index(colname) method. We can use it to access particular Pandas $\textit{rows}$, which are also Series objects:</p>

In [10]:
print(data)

# Grab the first row of data-- ny the index of the rox
my_row = data.loc[0]
print(f'\n Location of the 3rd row in the data: \n{my_row}')

# Getting multiple values from a df.loc
multiple_peeps = data.loc[  [0,1,2]  ] ['names']
print(f'\n Location of multiple people: {multiple_peeps}')
print(type(multiple_peeps))


# Using df.loc and producing a DataFrame at the end
multpiple_peeps_df = data.loc[[0,1,2]][['names', 'ages']]
# Reset our my_people dataframe index
reset_index = data.loc[ [0,1,2,3] ].set_index('names')
print(reset_index.loc['Alice'])

     names  ages
0    Alice    19
1      Bob    31
2    James    28
3     Beth    18
4     John    30
5    Sally    34
6  Richard    22
7   Lauren    19
8  Brandon    20
9  Sabrina    33

 Location of the 3rd row in the data: 
names    Alice
ages        19
Name: 0, dtype: object

 Location of multiple people: 0    Alice
1      Bob
2    James
Name: names, dtype: object
<class 'pandas.core.series.Series'>
ages    19
Name: Alice, dtype: int32


##### keys()

In [11]:
# Access all of the keys/columns of the dataframe
# Dataframe.keys()

data.keys()
print(type(bos17.keys()))
print(type(bos17.keys().tolist()))


<class 'pandas.core.indexes.base.Index'>
<class 'list'>


##### Slicing a DataFrame

In [12]:
# printing all data for context
data[2:5]

names  ages
2  James    28
3   Beth    18
4   John    30

### Built-In Methods <br>

<p>These are methods that are frequently used when using Pandas to make your life easier. It is possible to spend a whole week simply exploring the built-in functions supported by DataFrames in Pandas. Here however, we will simply highlight a few ones that might be useful, to give you an idea of what's possible out of the box with Pandas:</p>

##### head()

In [13]:
# DataFrame.head()  -- Accepts integer parameter(gives access to more rows)
bos17.head()

Rk Pos               Name  Age    G   PA   AB   R    H  2B  ...    OBP  \
0   1   C  Christian Vazquez   26   99  345  324  43   94  18  ...  0.330   
1   2  1B     Mitch Moreland   31  149  576  508  73  125  34  ...  0.326   
2   3  2B     Dustin Pedroia   33  105  463  406  46  119  19  ...  0.369   
3   4  SS    Xander Bogaerts   24  148  635  571  94  156  32  ...  0.343   
4   5  3B      Rafael Devers   20   58  240  222  34   63  14  ...  0.338   

     SLG    OPS  OPS+   TB  GDP  HBP  SH  SF  IBB  
0  0.404  0.735    91  131   14    3   0   1    0  
1  0.443  0.769    99  225   14    6   0   5    6  
2  0.392  0.760   100  159   11    2   2   4    4  
3  0.403  0.746    95  230   17    6   0   2    6  
4  0.482  0.819   111  107    5    0   0   0    3  

[5 rows x 28 columns]

##### tail()

In [14]:
# DataFrame.tail()  -- Accepts integer parameter(gives access to more rows)
bos17.tail()

Rk Pos             Name  Age   G  PA  AB  R   H  2B  ...    OBP    SLG  \
18  19  IF  Marco Hernandez   24  21  60  58  7  16   3  ...  0.300  0.328   
19  20  UT      Rajai Davis   36  17  38  36  7   9   2  ...  0.289  0.306   
20  21  UT     Steve Selsky   27   8   9   9  0   1   1  ...  0.111  0.222   
21  22  UT    Blake Swihart   25   6   7   5  1   1   0  ...  0.429  0.200   
22  23  2B   Chase d'Arnaud   30   2   1   1  2   1   0  ...  1.000  1.000   

      OPS  OPS+  TB  GDP  HBP  SH  SF  IBB  
18  0.628    65  19    0    1   0   0    0  
19  0.595    56  11    2    1   0   0    0  
20  0.333   -16   2    0    0   0   0    0  
21  0.629    74   1    0    0   0   0    0  
22  2.000   428   1    0    0   0   0    0  

[5 rows x 28 columns]

##### shape

In [15]:
# The dataframe has a shape property, just like a NumPy matrix. 
# print(df.shape) -- DataFrame.shape -- No Parameter

bos17.shape



(23, 28)

##### describe() <br>
<p>Probably one of the most important methods to understand...</p>

In [16]:
# Collect summary statistics in one line
# DataFrame.describe() -- Accepts parameters (include, exclude)
bos18.describe(include = 'all')

Rk  Pos        Name        Age           G          PA  \
count   20.00000   20          20  20.000000   20.000000   20.000000   
unique       NaN   11          20        NaN         NaN         NaN   
top          NaN   UT  Sandy Leon        NaN         NaN         NaN   
freq         NaN    4           1        NaN         NaN         NaN   
mean    10.50000  NaN         NaN  29.100000   82.350000  314.100000   
std      5.91608  NaN         NaN   4.666792   52.916294  230.972065   
min      1.00000  NaN         NaN  21.000000    2.000000    7.000000   
25%      5.75000  NaN         NaN  25.000000   37.000000  125.500000   
50%     10.50000  NaN         NaN  29.500000   85.500000  278.500000   
75%     15.25000  NaN         NaN  32.500000  129.250000  510.250000   
max     20.00000  NaN         NaN  37.000000  150.000000  661.000000   

                AB           R           H         2B  ...        OBP  \
count    20.000000   20.000000   20.000000  20.000000  ...  20.000000   
unique         NaN         NaN         NaN        NaN  ...        NaN   
top            NaN         NaN         NaN        NaN  ...        NaN   
freq           NaN         NaN         NaN        NaN  ...        NaN   
mean    280.200000   43.600000   75.250000  17.650000  ...   0.307700   
std     203.871631   38.293053   62.462515  15.624795  ...   0.068843   
min       6.000000    0.000000    1.000000   0.000000  ...   0.143000   
25%     115.250000   16.500000   28.000000   6.000000  ...   0.262000   
50%     258.000000   29.000000   49.500000  11.000000  ...   0.305500   
75%     475.500000   62.250000  115.000000  26.250000  ...   0.360500   
max     579.000000  129.000000  188.000000  47.000000  ...   0.438000   

              SLG        OPS        OPS+          TB        GDP        HBP  \
count   20.000000  20.000000   20.000000   20.000000  20.000000  20.000000   
unique        NaN        NaN         NaN         NaN        NaN        NaN   
top           NaN        NaN         NaN         NaN        NaN        NaN   
freq          NaN        NaN         NaN         NaN        NaN        NaN   
mean     0.387500   0.695200   86.450000  127.200000   6.500000   2.750000   
std      0.135884   0.200419   52.199491  113.226183   5.633546   3.274704   
min      0.091000   0.310000  -17.000000    1.000000   0.000000   0.000000   
25%      0.304000   0.588000   59.500000   37.500000   1.000000   0.000000   
50%      0.399000   0.712500   91.000000   72.500000   5.500000   2.000000   
75%      0.441000   0.788000  112.500000  192.000000   9.000000   4.250000   
max      0.640000   1.078000  186.000000  358.000000  19.000000  11.000000   

               SH         SF        IBB  
count   20.000000  20.000000  20.000000  
unique        NaN        NaN        NaN  
top           NaN        NaN        NaN  
freq          NaN        NaN        NaN  
mean     0.350000   2.400000   1.900000  
std      0.812728   2.436564   3.110255  
min      0.000000   0.000000   0.000000  
25%      0.000000   0.000000   0.000000  
50%      0.000000   2.000000   0.000000  
75%      0.000000   4.250000   2.250000  
max      3.000000   7.000000  11.000000  

[11 rows x 28 columns]

##### sort_values()

In [17]:
# Sort based on many labels, with left-to-right priority
# sorted_data = data.sort_values('ages').reset_index()

# DataFrame.sort_values('key')
sorted_data = data.sort_values('ages', kind = 'mergesort')

sorted_data.reset_index(drop=True)


names  ages
0     Beth    18
1    Alice    19
2   Lauren    19
3  Brandon    20
4  Richard    22
5    James    28
6     John    30
7      Bob    31
8  Sabrina    33
9    Sally    34

##### .columns

In [18]:
# will show all cols headers
# DataFrame.columns -- has no parameters

print(bos17.columns)
print(f'\n The type of bos17.columns: {type(bos17.columns)}')

# Keys brings back the 'index' of whatever data types we are working with
print(bos17.keys())
print(f'\n The type of bos17.keys(): {type(bos17.keys())}')

# Keys brings back the 'index' of whatever data type we are working with 


Index(['Rk', 'Pos', 'Name', 'Age', 'G', 'PA', 'AB', 'R', 'H', '2B', '3B', 'HR',
       'RBI', 'SB', 'CS', 'BB', 'SO', 'BA', 'OBP', 'SLG', 'OPS', 'OPS+', 'TB',
       'GDP', 'HBP', 'SH', 'SF', 'IBB'],
      dtype='object')

 The type of bos17.columns: <class 'pandas.core.indexes.base.Index'>
Index(['Rk', 'Pos', 'Name', 'Age', 'G', 'PA', 'AB', 'R', 'H', '2B', '3B', 'HR',
       'RBI', 'SB', 'CS', 'BB', 'SO', 'BA', 'OBP', 'SLG', 'OPS', 'OPS+', 'TB',
       'GDP', 'HBP', 'SH', 'SF', 'IBB'],
      dtype='object')

 The type of bos17.keys(): <class 'pandas.core.indexes.base.Index'>


### In-Class Exercise #2 - Describe & Sort Boston Red Sox Hitting Data <br>
<p>Take the data that you read in earlier from the Red Sox csv's and use the describe method to understand the data better. Compare the two years and decide which team is having the better year. Then sort the values based on Batting Average.</p>

In [19]:

bos17.describe(include = 'all')
bos18.describe(include = 'all')


# COmpare the summary statistics
# For simplicity, lets compare the mean batting average for each year
mean_batting_average_2017 = bos17['BA'].mean()
mean_batting_average_2018 = bos18['BA'].mean()

print("Mean Batting Average for 2017:", mean_batting_average_2017)
print("Mean batting Average for 2018:", mean_batting_average_2018)

# Decide which year had the better performance
if mean_batting_average_2017 > mean_batting_average_2018:
    print("2017 was a better year for the Red Sox in terms of batting average.")
elif mean_batting_average_2017 < mean_batting_average_2018:
    print("2018 was a better year for the Red Sox in terms of batting average.")
else:
    print("Both years had the same mean batting average.")
    
# Sort the values based on Batting Average
sorted_2017 = bos17.sort_values(by='BA', ascending =False)
sorted_2018 = bos18.sort_values(by='BA', ascending=False)

# Display the sorted values
print("Sorted values based on Batting Average for 2017:")
print(sorted_2017)
print("\nSorted values based on Batting Average for 2018:")
print(sorted_2018)

Mean Batting Average for 2017: 0.2784347826086956
Mean batting Average for 2018: 0.23795000000000002
2017 was a better year for the Red Sox in terms of batting average.
Sorted values based on Batting Average for 2017:
    Rk Pos                Name  Age    G   PA   AB    R    H  2B  ...    OBP  \
22  23  2B      Chase d'Arnaud   30    2    1    1    2    1   0  ...  1.000   
12  13  2B       Eduardo Nunez   30   38  173  165   23   53  12  ...  0.353   
2    3  2B      Dustin Pedroia   33  105  463  406   46  119  19  ...  0.369   
0    1   C   Christian Vazquez   26   99  345  324   43   94  18  ...  0.330   
4    5  3B       Rafael Devers   20   58  240  222   34   63  14  ...  0.338   
18  19  IF     Marco Hernandez   24   21   60   58    7   16   3  ...  0.300   
3    4  SS     Xander Bogaerts   24  148  635  571   94  156  32  ...  0.343   
5    6  LF   Andrew Benintendi   22  151  658  573   84  155  26  ...  0.352   
17  18  UT         Tzu-Wei Lin   23   25   66   56    7   15  

### Filtration <br>
<p>Let's look at how to filter dataframes for rows that fulfill a specific conditon.</p>

##### Conditionals

In [20]:
# Conditional boolean dataframe
data['ages'] >= 25

0    False
1     True
2     True
3    False
4     True
5     True
6    False
7    False
8    False
9     True
Name: ages, dtype: bool

##### Subsetting

In [21]:
# exactly like numpy
data[data['ages'] >= 25]

names  ages
1      Bob    31
2    James    28
4     John    30
5    Sally    34
9  Sabrina    33

### Column Transformations <br>
<p>Rarely, if ever, will the columns in the original raw dataframe read from CSV or database table be the ones you actually need for your analysis. You will spend lots of time constantly transforming columns or groups of columns using general computational operations to produce new ones that are functions of the old ones. Pandas has full support for this: Consider the following dataframe containing membership term and renewal number for a group of customers:</p>

In [25]:
# Generate some fake data
customer_id = np.random.randint(1000, 1100, 10)
renewal_nbr = np.random.randint(0, 10, 10)
customer_dict = {1: 0.5, 0: 1}
terms_in_years =[customer_dict[key] for key in np.random.randint(0,2,10)]

random_data = {
    'customer_id': customer_id,
    'renewal_nbr': renewal_nbr,
    'term_in_years': terms_in_years
}

customers = pd.DataFrame.from_dict(random_data)
customers

customer_id  renewal_nbr  term_in_years
0         1032            1            1.0
1         1039            4            0.5
2         1034            9            1.0
3         1096            1            0.5
4         1060            2            0.5
5         1098            3            1.0
6         1099            1            1.0
7         1063            0            0.5
8         1098            7            1.0
9         1075            5            1.0

##### Generating a New Column w/Data

In [26]:
# DataFrame['key'] = Some Calculation from our DataFrame Columns
customers['customer_tenure'] = customers['renewal_nbr'] * customers['term_in_years']
customers

customer_id  renewal_nbr  term_in_years  customer_tenure
0         1032            1            1.0              1.0
1         1039            4            0.5              2.0
2         1034            9            1.0              9.0
3         1096            1            0.5              0.5
4         1060            2            0.5              1.0
5         1098            3            1.0              3.0
6         1099            1            1.0              1.0
7         1063            0            0.5              0.0
8         1098            7            1.0              7.0
9         1075            5            1.0              5.0

##### User Defined Function

<p>If what you want to do to a column that can't be represented by simple mathematical operations, you can write your own $\textit{user defined function}$ with the full customizability available in Python and any external Python packages, then map it directly onto a column. Let's add some ages to our customer dataframe, and then classify them into our custom defined grouping scheme:</p>

In [27]:
# use .apply to map over dataframe

# Create a new column for ages
customers['ages'] = np.random.randint(18,70,10)

# User defined function
def make_age_groups(age):
    if 10 <= age and 20 > age:
        return 'Teenager'
    elif 20 <= age and 35 > age:
        return 'Young Adult'
    elif 35 <= age and 65 > age:
        return 'Adult'
    elif 65 <= age:
        return 'Senior'
    
# print(make_age_groups)
customers['age_group'] = customers['ages'].apply(make_age_groups)
customers

customer_id  renewal_nbr  term_in_years  customer_tenure  ages    age_group
0         1032            1            1.0              1.0    27  Young Adult
1         1039            4            0.5              2.0    46        Adult
2         1034            9            1.0              9.0    57        Adult
3         1096            1            0.5              0.5    67       Senior
4         1060            2            0.5              1.0    62        Adult
5         1098            3            1.0              3.0    56        Adult
6         1099            1            1.0              1.0    52        Adult
7         1063            0            0.5              0.0    20  Young Adult
8         1098            7            1.0              7.0    42        Adult
9         1075            5            1.0              5.0    45        Adult

<p>As a last example I'll show here how you would use a lambda function to create a UDF that depends on $\textit{more than one}$ column:</p>

<li>UDF = User Defined Function</li>

In [28]:
#Axis for apply can only be 1 or 0 -- 1 being the X axis 0 being the Y axis

def make_loyalty_age_group(row):
    age = row['ages']
    tenure = row['customer_tenure']
    
    if 10 <= age and 20 > age:
        age_group = 'Teenager'
    elif 20 <= age and 35 > age:
        age_group = 'Young Adult'
    elif 35 <= age and 65 > age:
        age_group = 'Adult'
    elif 65 <= age:
        age_group = 'Senior'
    
    if tenure > 2.0:
        loyalty_age_group = f'Loyal {age_group}'
    else:
        loyalty_age_group = f'New {age_group}'
        
    return loyalty_age_group
    
customers['loyalty_age_group'] = customers.apply(make_loyalty_age_group, axis = 1)
customers

customer_id  renewal_nbr  term_in_years  customer_tenure  ages  \
0         1032            1            1.0              1.0    27   
1         1039            4            0.5              2.0    46   
2         1034            9            1.0              9.0    57   
3         1096            1            0.5              0.5    67   
4         1060            2            0.5              1.0    62   
5         1098            3            1.0              3.0    56   
6         1099            1            1.0              1.0    52   
7         1063            0            0.5              0.0    20   
8         1098            7            1.0              7.0    42   
9         1075            5            1.0              5.0    45   

     age_group loyalty_age_group  
0  Young Adult   New Young Adult  
1        Adult         New Adult  
2        Adult       Loyal Adult  
3       Senior        New Senior  
4        Adult         New Adult  
5        Adult       Loyal Adult  
6        Adult         New Adult  
7  Young Adult   New Young Adult  
8        Adult       Loyal Adult  
9        Adult       Loyal Adult

### In-Class Exercise #3 - Create Your Own UDF <br>
<p>Using the Boston Red Sox data, create your own UDF which creates a new column called 'All-Star' and puts every player with either a batting average over .280 or an on base percentage of over .360 with a result of 'Yes' in the column and 'No' if not.</p>

In [29]:
def is_all_star(row):
    if row['BA'] > 0.280 or row['OBP'] > 0.360:
        return 'Yes'
    else:
        return 'No'
    
    
# Read the data from the CSV file
bos17 = pd.read_csv('redsox_2017_hitting.txt')
bos18 = pd.read_csv('redsox_2018_hitting.txt')

bos17['AllStar'] = bos17.apply(is_all_star, axis=1)
bos18['AllStar'] = bos18.apply(is_all_star, axis=1)

print(bos17)
print(bos18)


    Rk Pos                Name  Age    G   PA   AB    R    H  2B  ...    SLG  \
0    1   C   Christian Vazquez   26   99  345  324   43   94  18  ...  0.404   
1    2  1B      Mitch Moreland   31  149  576  508   73  125  34  ...  0.443   
2    3  2B      Dustin Pedroia   33  105  463  406   46  119  19  ...  0.392   
3    4  SS     Xander Bogaerts   24  148  635  571   94  156  32  ...  0.403   
4    5  3B       Rafael Devers   20   58  240  222   34   63  14  ...  0.482   
5    6  LF   Andrew Benintendi   22  151  658  573   84  155  26  ...  0.424   
6    7  CF  Jackie Bradley Jr.   27  133  541  482   58  118  19  ...  0.402   
7    8  RF        Mookie Betts   24  153  712  628  101  166  46  ...  0.459   
8    9  DH      Hanley Ramirez   33  133  553  496   58  120  24  ...  0.429   
9   10   C          Sandy Leon   28   85  301  271   32   61  14  ...  0.354   
10  11  UT         Chris Young   33   90  276  243   30   57  12  ...  0.387   
11  12  3B       Deven Marrero   26   71

### Aggregations <br>
<p>The raw data plus some transformations is generally only half the story. Your objective is to extract actual insights and actionable conclusions from the data, and that means reducing it from potentially billions of rows to some summary statistics via aggregation functions.</p>

##### groupby() <br>
<p>The .groupby() function is in some ways a 'master' aggregation.</p> 

<p>Data tables will usually reserve one column as a primary key - that is, a column for which each row has a unique value. This is to facilitate access to the exact rows of a data table that a user wants to view. The other columns will often have repeated values, such as the age groups in the above examples. We can use these columns to explore the data using the Pandas API:</p>

In [30]:
# also introducing .count() here, exact same as to how it's used in SQL

# Using the groupby with the column intact as a column/key
# customers.groupby('age_group', as_index = False).count()[['customers_id','age_group']]

customers.groupby('age_group').count()

customer_id  renewal_nbr  term_in_years  customer_tenure  ages  \
age_group                                                                     
Adult                  7            7              7                7     7   
Senior                 1            1              1                1     1   
Young Adult            2            2              2                2     2   

             loyalty_age_group  
age_group                       
Adult                        7  
Senior                       1  
Young Adult                  2

##### Type of groupby()

<p>The result is a new dataframe, the columns of which all contain the counts of the grouped field. Notice the type of a grouped dataframe:</p>

In [34]:
print(type(customers), 'Regular Data Frame')

<class 'pandas.core.frame.DataFrame'> Regular Data Frame


<p>This is because simply grouping data doesn't quite make sense without an aggregation function like count() to pair with. In this case, we're counting occurances of the grouped field, but that's not all we can do. We can take averages, standard deviations, mins, maxes and much more! Let's see how this works a bit more:</p>

##### mean()

In [35]:
# mean = average
customers.groupby('age_group').mean()[['customer_tenure',  'ages']]

TypeError: Could not convert New AdultLoyal AdultNew AdultLoyal AdultNew AdultLoyal AdultLoyal Adult to numeric

##### groupby() w/Multiple Columns

<p>We end up with the average age of the groups in the last column, the average tenure in the tenure column, and so on and so forth. You can even split the groups more finely by passing a list of columns to group by:</p>

In [36]:
customers.groupby(['age_group', 'ages']).mean().sort_values('ages', ascending = False)

TypeError: Could not convert Loyal Adult to numeric

##### drop_duplicates()

<p>Drops all duplicates from the current dataframe</p>

In [37]:
customers

customer_id  renewal_nbr  term_in_years  customer_tenure  ages  \
0         1032            1            1.0              1.0    27   
1         1039            4            0.5              2.0    46   
2         1034            9            1.0              9.0    57   
3         1096            1            0.5              0.5    67   
4         1060            2            0.5              1.0    62   
5         1098            3            1.0              3.0    56   
6         1099            1            1.0              1.0    52   
7         1063            0            0.5              0.0    20   
8         1098            7            1.0              7.0    42   
9         1075            5            1.0              5.0    45   

     age_group loyalty_age_group  
0  Young Adult   New Young Adult  
1        Adult         New Adult  
2        Adult       Loyal Adult  
3       Senior        New Senior  
4        Adult         New Adult  
5        Adult       Loyal Adult  
6        Adult         New Adult  
7  Young Adult   New Young Adult  
8        Adult       Loyal Adult  
9        Adult       Loyal Adult

In [38]:
customers_copy = customers.drop_duplicates('renewal_nbr').reset_index(drop=True)
customers_copy

customer_id  renewal_nbr  term_in_years  customer_tenure  ages  \
0         1032            1            1.0              1.0    27   
1         1039            4            0.5              2.0    46   
2         1034            9            1.0              9.0    57   
3         1060            2            0.5              1.0    62   
4         1098            3            1.0              3.0    56   
5         1063            0            0.5              0.0    20   
6         1098            7            1.0              7.0    42   
7         1075            5            1.0              5.0    45   

     age_group loyalty_age_group  
0  Young Adult   New Young Adult  
1        Adult         New Adult  
2        Adult       Loyal Adult  
3        Adult         New Adult  
4        Adult       Loyal Adult  
5  Young Adult   New Young Adult  
6        Adult       Loyal Adult  
7        Adult       Loyal Adult

In [40]:
# Send customer data into a CSV file
customers_copy.to_csv('../Downloads/customers.csv')

<p>Thus the groupby operation allows you to rapidly make summary observations about the state of your entire dataset at flexible granularity. In one line above, we actually did something very complicated - that's the power of the dataframe. In fact, the process often consists of several iterative groupby operations, each revealing greater insight than the last - if you don't know where to start with a dataset, try a bunch of groupbys!</p>

### Homework Excersise #1 - Find the Total Number of Runs and RBIs for the Red Sox <br>
<p>Get total number of home runs and rbi's</p>

In [57]:
# step 1: Add a new column with the key 'Team' and all column values should be 'BOS'

# step 2: Group by the 'Team' column and get total home runs and rbi's

# Produce data for both 2017 and 2018 (ie print both seperated by a newline character \n)

# Read data from both files, specifying separator as comma
bos17 = pd.read_csv('redsox_2017_hitting.txt', sep=',')
bos18 = pd.read_csv('redsox_2018_hitting.txt', sep=',')

# Add a new column with the key 'Team' and all column values should be 'BOS'
bos17['Team'] = 'BOS'
bos18['Team'] = 'BOS'

# Calculate total home runs and RBIs for 2017
total_hr_17 = bos17['HR'].sum()
total_rbi_17 = bos17['RBI'].sum()

# Calculate total home runs and RBIs for 2018
total_hr_18 = bos18['HR'].sum()
total_rbi_18 = bos18['RBI'].sum()

# Print the results
print("2017:")
print("Total Home Runs:", total_hr_17)
print("Total RBIs:", total_rbi_17)
print("\n")
print("2018:")
print("Total Home Runs:", total_hr_18)
print("Total RBIs:", total_rbi_18)


2017:
Total Home Runs: 168
Total RBIs: 735


2018:
Total Home Runs: 208
Total RBIs: 826


## Find the average age of runners in the 2017 Boston Marathon

In [58]:
marathon = pd.read_csv("boston_marathon2017_edited.csv")

# Calculate the average age of runners
average_age = marathon['Age'].mean()

print("Average age of runners in the 2017 Boston Marathon:", average_age)

Average age of runners in the 2017 Boston Marathon: 42.587731919727375
